threatgrid/ctia
# Wait_for on update / patch / delete

This PR enables to use the `wait_for` query parameter for not only Creation of resources, but also update, patch, delete, bulk.

> Close #https://github.com/threatgrid/iroh/issues/2957
> Related #https://github.com/threatgrid/iroh/issues/2672

In order to enable to pass the wait_for param from routes to store, this PR modifies the protocal IStore with a new `params` parameter for `update-record` and `delete-record` functions. 

<a name="qa">[§](#qa)</a> QA
============================

- Check UPDATE / PATCH
  - Perform a  loop that repeats the following test 100 times: 
  1. Update/Patch any entity with `wait_for=true` as query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4. Check that the resource is well updated/patched

  - Perform a loop that repeats the following test 100 times: 
  1. Update/Patch any entity with`wait_for=false` as query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4. Check the returned entity: most should not be updated/patched (in some lucky cases some can be)

  - Perform a loop that repeats the following test 100 times: 
  1. Update/Patch any entity without the `wait_for` query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4. Check the returned entity: most should not be updated/patched (in some lucky cases some can be)


- Check DELETE
  - Perform a  loop that repeats the following test 100 times: 
  1. Delete any entity with `wait_for=true` as query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4. Check that the response status is 404

  - Perform a loop that repeats the following test 100 times: 
  1. Delete any entity with`wait_for=false` as query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4. Check the returned status: most of the status of the GET should be 200 (in some lucky cases some can be 404)

  - Perform a loop that repeats the following test 100 times: 
  1. Delete any entity without the `wait_for` query param
  2. Retrieve the response ID and generate the short version of that ID (last part of the the url)
  3. Retrieve that resource using the short ID
  4.  Check that the response status is 404

- Check Bulk
  - Perform a  loop that repeats the following test 100 times: 
  1. Bulk import entities with `wait_for=true` as query param
  2. Retrieve reponse Ids and generate the short version of these IDs (last part of the the urls)
  3. Retrieve the resources using these short IDs
  4. Check that the response statuses are 200

  - Perform a loop that repeats the following test 100 times: 
  1. Bulk import entities with `wait_for=true` as query param
  2. Retrieve reponse Ids and generate the short version of these IDs (last part of the the urls)
  3. Retrieve the resources using these short IDs
  4. Check the returned status: most of the status of the GET should be 404 (in some lucky cases some can be 200)

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: When updating / deleting a resource or bulk / bundle importing, the query param `wait_for=true` could be used to deal with inconsistencies.
```
